addon/authenticators/jwt.js

```diff
@@ -101,34 +101,46 @@ export default TokenAuthenticator.extend({
   restore: function(data) {
     var _this = this,
       dataObject = Ember.Object.create(data);
-    
+
     return new Ember.RSVP.Promise(function(resolve, reject) {
       var now = (new Date()).getTime(),
         expiresAt = _this.resolveTime(dataObject.get('expiresAt')),
         token = dataObject.get(_this.tokenPropertyName);
+      if (Ember.isEmpty(token)) {
+        return reject(new Error('empty token'));
+      }
 
-      if (!Ember.isEmpty(expiresAt) && !Ember.isEmpty(token) && expiresAt > now) {
-        if (_this.refreshAccessTokens) {
-          _this.refreshAccessToken(data.token).then(function(data) {
-            resolve(data);
-          }, reject);
-        } else {
-          reject();
+       if (Ember.isEmpty(expiresAt)) {
+        // Fetch the expire time from the token data since `expiresAt`
+        // wasn't included in the data object that was passed in.
+        var tokenData = _this.getTokenData(data[_this.tokenPropertyName]);
+        expiresAt = _this.resolveTime(tokenData[_this.tokenExpireName]);
+        if (Ember.isEmpty(expiresAt)) {
+          return resolve(data);
         }
-      } else {
-        if (Ember.isEmpty(token)) {
-          reject();
-        } else {
-          // Fetch the expire time from the token data since `expiresAt`
-          // wasn't included in the data object that was passed in.
-          var tokenData = _this.getTokenData(data[_this.tokenPropertyName]),
-            tokenExpiresAt = tokenData[_this.tokenExpireName];
-
-          _this.scheduleAccessTokenRefresh(tokenExpiresAt, token);
 
+      }
+      if (expiresAt !== expiresAt) {
+        return reject(new Error('invalid expiration'));
+      }
+      if (expiresAt > now) {
+        var wait = expiresAt - now - (_this.refreshLeeway * 1000);
+        if (wait > 0) {
+          if (_this.refreshAccessTokens) {
+            _this.scheduleAccessTokenRefresh(expiresAt, token);
+          }
           resolve(data);
+        } else if (_this.refreshAccessTokens) {
+          resolve(_this.refreshAccessToken(data.token).then(function () {
+            return data;
```

Comment:
inside of a promise callback returning a value causes the promise to resolve into that value, so the fact we passed it to resolve in 134 is engouth